### PR TITLE
[ENH] Enhancements to encode_categorical method

### LINKED ISSUE #677: "as_categorical"
Another idea!

```python
@pf.register_dataframe_method
def as_categorical(df, column_name, categories=None, ordered=True):
    """
    Convert one column in the dataframe to a categorical one.
    """
    if categories is None:
        categories = sorted(df[column_name].unique())
    df[column_name] = pd.Categorical(
        df[column_name], categories=categories, ordered=ordered,
    )
    return df
```

It replaces repeatedly doing this piece:

```python
df[column_name] = pd.Categorical(
    df[column_name], categories=categories, ordered=ordered,
)
```

Leaving this here for now.

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Update to encode_categorical method
- Option to specify categories and order for the categorical columns.

**This PR resolves #677  .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [ ] Build the docs locally.
- [ ] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by samukweku** [comment]
@ericmjl @hectormz Could you please have a look at this (in between baby time :smiley: ) and possibly show me how to resolve this error from the `test_convert_excel_date` file?

**Comment by samukweku** [comment]
@ericmjl @hectormz so I found the error. It has to do with `xlrd`, which is no longer maintained. One option is to change the argument engine to `openpyxl`, which still has maintainers. I will create a PR and openpyxl as a dependency in the `environment-dev.yml` file.

**Comment by hectormz** [comment]
Hi @samukweku I'm just looking at this now. So if I understand this PR and issue, that test theoretically shouldn't have been altered, right? You didn't originally touch anything that should have made it fail?

It works fine on my machine using pandas 1.0.5 and the latest 1.1.4. So I wonder if there is something else at play. (I checked this PR out locally)

**Comment by hectormz** [comment]
I believe the issue stems from the fact that the most recent PR to pass #770 used python 3.8 in the CI, and now it is using python 3.9, and there are some issues with things being removed, etc.

So the "fix" for us right now, may be to not support 3.9 until other things change.

**Comment by samukweku** [comment]
Thanks @hectormz . Do you mind showing me how you deciphered the python version, and what can be done to prevent python3.9?

**Comment by samukweku** [comment]
@hectormz I can see that it now reads python3.9, even though the environment_dev is 3.7. any suggestions on how to turn off python3.9?

**Comment by samukweku** [comment]
@hectormz @ericmjl thanks! apparently` xlrd` has an issue with xlsx files hence the warnings and finally the errors that popped up since the CI somehow uses the latest python version (3.9). setting the engine as `openpyxl` resolves the issue.

**Comment by hectormz** [comment]
@samukweku  I added a (test) PR #774 to verify that this is the issue. I wasn't aware that the CI is just using the `environment.yml` file instead of python version specified in `pipeline-master.yml`. I imagine `pandas` would have their own default fix to this, so I think it would be fine to restrict running tests to <3.9 until they have resolved that issue.

File: environment-dev.yml
Comment (hectormz) [inline environment-dev.yml]:
```suggestion
- xorg-libxrender

```

**Review by hectormz — CHANGES_REQUESTED** [review]

**Comment by hectormz** [comment]
@samukweku Our suspicions about python 3.9 seem to be correct. Changing the `environment.yml` in my test PR and the deprecation warnings/errors show it's a 3.9 problem. HOWEVER, when I installed python 3.9 and installed `pandas` & `xlrd`, I had no issues opening the file in `pandas`. So I have feeling there is actually some weird `conda` problem going on, or maybe something with CI machine?


But I couldn't even install `pyjanitor` on python 3.9 on Windows because it would get hung up on `scikit-learn`, so maybe we really shouldn't be python 3.9 compatible/approved at the moment.

**Comment by samukweku** [comment]
@hectormz , totally agree that we should set it for python 3.8 or 3.7.  Thanks for the insight. Learnt a few more things about infrastructure, and working in a team.

**Comment by ericmjl** [comment]
@hectormz I trust your review here, thanks for doing a thorough job on this one. Off to the paediatrician today; let's see if we can get in one more review before the start of next week, otherwise, please do the honours of merging :smile:. As always, @samukweku thanks as well for taking this PR on; a cursory glance through the code and I see the wonderful amount of detail in the docs, which is amazing!

**Comment by samukweku** [comment]
@ericmjl @hectormz I put a pause on this because I just discovered that there is an [encode_categorical](https://pyjanitor.readthedocs.io/reference/janitor.functions/janitor.encode_categorical.html#janitor.encode_categorical) function in pyjanitor that create categorical columns. It only creates category dtype; you cannot set order to True or pass categories. Need suggestions from the team on what steps to take

**Comment by ericmjl** [comment]
@samukweku you're right, in my brain mushiness, I totally forgot that `encode_categorical` existed.

It seems like you've got some enhancements in here that could go into `encode_categorical` instead, what do you think of putting them in there? For example, the ordering is one that I think is missing from `encode_categorical`.

**Comment by samukweku** [comment]
Cool! I will move the PR over to `encode_categorical`. I have an idea on using kwargs with a namedtuple, to significantly reduce the lengthy checks and code that I wrote. fingers crossed it actually works.

**Review by ericmjl — APPROVED** [review]
@samukweku I approve! Please do the honours :smile:.

**Comment by ericmjl** [comment]
@hectormz, only your review is left blocking! If you've got time, would you be kind enough to leave a note on what you think about it?

**Review by hectormz — APPROVED** [review]

**Comment by ericmjl** [comment]
I finally got around to merging the PR. Thank you, @samukweku and @hectormz!!!

**Pull request merged**
